### PR TITLE
[Master] - Slice 626127: [Excise Tax][VENDOR] Multiple Excise Taxes per Item

### PR DESCRIPTION
**Pull request by v-rohangarg20**
Workitem 626127: [Master] - Slice 626127: [Excise Tax][VENDOR] Multiple Excise Taxes per Item

Fixes [AB#626127](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/626127)

**Issue:** An item could be linked to only a single excise tax type, because the excise tax setup (Excise Tax Type, Quantity for Excise Tax, Excise Unit of Measure Code) lived as single-value fields directly on the Item table. Vendors that need more than one excise tax on the same item could not be modeled, and excise journal generation only ever produced lines for that one tax type.

**Cause:** The excise tax configuration was stored as fields on the Item record (one value each) and the journal-generation logic (Excise Tax Calculation) iterated Items filtered by the single "Excise Tax Type" field, so the data model physically allowed only one excise tax per item.

**Solution:** Introduced a dedicated "Item Excise Tax" table (7415) keyed on Item No. + Excise Tax Type Code, exposed via the "Item Excise Taxes" list page (7416, opened from the Item Card with a Copy from Item action) and an "Item Excise Tax" API page (7417). Excise journal generation now iterates this table and became tax-type aware (per-tax-type existence/posted checks, plus AllExciseTaxesPostedForItemLedgerEntry so the ILE "Excise Tax Posted" flag is set only once every applicable enabled/allowed tax type is posted), with new OnBefore/OnAfter insert integration events. The legacy Item excise fields are deprecated (ObsoleteState = Pending, tag 29.0, behind CLEAN29) and existing data is migrated to the new table by the "Excise Tax Upgrade" codeunit (7414). Permission sets and tests were updated accordingly.

### CODE REVIEW

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 1

# **Recommendation: Request Changes**

<details>
<summary><strong>What this PR does</strong></summary>

This PR moves item excise setup from single fields on Item to a new Item Excise Tax table, adds list/API pages, migrates existing item setup, and changes journal generation so one item ledger entry can produce one excise journal line per tax type. The main generation path is mostly coherent: it reads the new setup table and checks existing journal lines per tax type. However, posting still marks the item ledger entry as globally Excise Tax Posted, and generation still filters that Boolean before checking the tax type. That leaves a valid multi-tax flow broken when tax types are generated or posted in separate runs.
</details>

<details>
<summary><strong>Suggestions</strong></summary>

**S1** - *Posted flag blocks later tax types*
Do not use the single Item Ledger Entry.Excise Tax Posted flag as the gate for all item excise taxes. After one tax type is posted, UpdateItemLedgerEntryExciseTaxInfo sets the flag to true, and CreateExciseJournalLineForItem then skips the same ledger entry for every other tax type. Track posted state per tax type, or base the skip on the excise transaction log/journal line for the same ledger entry and tax type, and add a test where one tax type is posted before another is generated.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** This is financially sensitive journal generation. The new table, pages, API page, upgrade code, and obsolete item fields are broad but fit the Slice. The blocking risk is in src/Apps/W1/ExciseTaxes/app/src/codeunit/ExciseTaxCalculation.Codeunit.al and src/Apps/W1/ExciseTaxes/app/src/codeunit/ExciseTaxTransSubscriber.Codeunit.al: a partial or filtered posting can prevent later excise tax amounts from ever being generated for the same item ledger entry.

**Necessity:** The Slice is valid and important because one item can need more than one excise tax setup. The scope is appropriate for a feature increment, and tests cover generating two lines in one run. The missing posted-state scenario must be fixed before merge because it is a normal use of the existing tax type filter and posting flow.
</details>

---
> [AI-PR-REVIEW] version=1 system=github pr=9725 round=1 by=alexei-dobriansky at=2026-08-03T10:18:59Z lastSha=d0751f312d3a560dd504cd7a67768af32bc79490 suggestions=S1

File: src/Apps/W1/ExciseTaxes/app/src/page/ItemExciseTaxes.Page.al [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    layout
+    {
+        area(Content)
+        {
+            repeater(Setup)
+            {
+                field("Item No."; Rec."Item No.")
+                {
+                    ToolTip = 'Specifies the item number.';
```
Comment (ViditGupta1277) [inline src/Apps/W1/ExciseTaxes/app/src/page/ItemExciseTaxes.Page.al]: Add tooltip to the table

File: src/Apps/W1/ExciseTaxes/app/src/page/ItemExciseTaxes.Page.al [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                    Visible = false;
+                }
+                field("Excise Tax Type Code"; Rec."Excise Tax Type Code")
+                {
+                    ToolTip = 'Specifies the excise tax type that applies to this item.';
```
Comment (ViditGupta1277) [inline src/Apps/W1/ExciseTaxes/app/src/page/ItemExciseTaxes.Page.al]: ApplicationArea property missing on Page fields

**Review by alexei-dobriansky — CHANGES_REQUESTED** [review]
Please review the suggestion S1

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 2

# **Recommendation: Accept**

<details>
<summary><strong>What this PR does</strong></summary>

The commits since round 1 address the posted-state problem for multiple item excise taxes. Journal generation now checks the transaction log for the same item ledger entry and tax type, and the shared `Excise Tax Posted` flag is set only after every enabled tax type allowed for that entry type has been posted (`ExciseTaxCalculation.Codeunit.al:149-159,369-416`). The item-ledger update now runs after the transaction log insert (`ExciseTaxTransSubscriber.Codeunit.al:31-40`, `SustainabilityExcisePostMgt.Codeunit.al:37-38`), so the check uses the posted records. The new tests cover posting two tax types in separate runs and the disabled/not-allowed cases (`ExciseTaxCalculationTests.Codeunit.al:552-711`).
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Posted flag blocks later tax types | Addressed | No direct reply. Fixed in `b11f538b` and `05482639`, with focused regression tests. |
</details>

<details>
<summary><strong>New observations (commits since round 1)</strong></summary>

*None - the authored changes since round 1 address S1 and add the needed edge-case coverage.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** This remains financially sensitive journal generation. The new logic adds transaction-log lookups during generation and checks each configured item tax type during posting, but the work is bounded by the tax types configured for the item. The new post-insert event keeps the item-ledger update in the same posting transaction. No new correctness issue was found in the round-2 changes. The current GitHub workflow is still running and shows two failed unit-test jobs outside W1, so CI status must be resolved separately.

**Necessity:** The change is required because users can generate and post different excise tax types in separate runs. Without it, posting one type can permanently hide the same item ledger entry from the remaining types. The new code and tests now cover that normal filtered-posting flow.
</details>

---
> [AI-PR-REVIEW] version=1 system=github pr=9725 round=2 by=alexei-dobriansky at=2026-08-04T11:00:00Z lastSha=05482639c38ce22e4d80568b876a9377de1fb23a suggestions=S1:addressed parentRound=1

File: src/Apps/W1/ExciseTaxes/app/src/codeunit/ExciseTaxUpgrade.Codeunit.al [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        UpgradeTag.SetUpgradeTag(GetMultipleExciseTaxesPerItemUpgradeTag());
+    end;
+
+    local procedure MigrateItemExciseTaxSetup()
+    var
+        Item: Record Item;
+        ItemExciseTax: Record "Item Excise Tax";
+        ExciseTaxType: Record "Excise Tax Type";
+    begin
+#pragma warning disable AL0432
```
Comment (djukicmilica) [inline src/Apps/W1/ExciseTaxes/app/src/codeunit/ExciseTaxUpgrade.Codeunit.al:33]: why we need this pragma?

File: src/Apps/W1/ExciseTaxes/app/src/codeunit/ExciseTaxUpgrade.Codeunit.al [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        Item.SetLoadFields("Excise Tax Type", "Quantity for Excise Tax", "Excise Unit of Measure Code");
+        Item.SetFilter("Excise Tax Type", '<>%1', '');
+        if Item.FindSet() then
+            repeat
+                if not ItemExciseTax.Get(Item."No.", Item."Excise Tax Type") then begin
```
Comment (Rishukla90) [inline src/Apps/W1/ExciseTaxes/app/src/codeunit/ExciseTaxUpgrade.Codeunit.al]: move it to a procedure

**Review by Rishukla90 — APPROVED** [review]

File: src/Apps/W1/ExciseTaxes/app/src/codeunit/ExciseTaxUpgrade.Codeunit.al (hunk 4 above)
Comment (v-rohangarg20) [reply src/Apps/W1/ExciseTaxes/app/src/codeunit/ExciseTaxUpgrade.Codeunit.al]: Resolved. Thanks!

**Review by djukicmilica — DISMISSED** [review]

File: src/Apps/W1/ExciseTaxes/app/src/codeunit/ExciseTaxUpgrade.Codeunit.al (hunk 3 above)
Comment (v-rohangarg20) [reply src/Apps/W1/ExciseTaxes/app/src/codeunit/ExciseTaxUpgrade.Codeunit.al:33]: We’re obsoleting the field that’s showing the warning, so the pragma is used here to suppress it.

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 1

# **Recommendation: Accept with Suggestions**

<details>
<summary><strong>What this PR does</strong></summary>

This PR replaces the single excise-tax-type field on the `Item` table with a new `Item Excise Tax` table (7415) that allows one item to carry multiple excise tax types simultaneously. The three existing Item fields (`Excise Tax Type`, `Quantity for Excise Tax`, `Excise Unit of Measure Code`) are marked obsolete behind `#if not CLEAN29` and an upgrade codeunit migrates existing data to the new table.

The correctness is solid. `CreateExciseJournalLineForItem` now iterates `ItemExciseTax` records instead of the `Item` record, correctly producing one journal line per tax-type per item ledger entry. `ExciseJournalLineExist` was extended with a `TaxType` filter so the same ledger entry can produce lines for each type without being blocked by an earlier type's line. `AllExciseTaxesPostedForItemLedgerEntry` correctly skips disabled and non-permitted tax types when deciding whether to mark the ledger entry as fully posted, and correctly skips the current tax type's log lookup (the log entry is already in the transaction when the event fires). The move of `UpdateItemLedgerEntryExciseTaxInfo` to the new `OnAfterInsertExciseTaxesTransactionLog` event — separate from `UpdateFALedgerEntryExciseTaxInfo` — is logically clean. The test suite is comprehensive and covers the key scenarios including partial posting, disabled types, and non-permitted entry types.
</details>

<details>
<summary><strong>Suggestions</strong></summary>

**S1** - *Missing `SetLoadFields` before `ExciseTaxType.Get` in upgrade*
In `CreateItemExciseTaxFromItem`, the call `ExciseTaxType.Get(Item."Excise Tax Type")` loads all fields but only `Description` is used. Add `ExciseTaxType.SetLoadFields(Description)` before the `Get` call. This reduces the per-row read cost during migration, which matters when the upgrade processes large item catalogues.

**S2** - *Redundant enabled check duplicates the table relation in `ItemExciseTax` `OnValidate`*
The `TableRelation` on `"Excise Tax Type Code"` is already filtered to `where(Enabled = const(true))`, so the `if not ExciseTaxType.Enabled then Error(...)` block inside `OnValidate` can never fire through any normal or programmatic path that runs the relation check first. The `ExciseTaxType.Get()` call is still needed to read `Description`; only the redundant enabled check should be removed to keep the code clean.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The change touches the posting path for excise tax journal lines (`ExciseTaxCalculation`, `ExciseTaxTransSubscriber`, `SustainabilityExcisePostMgt`). The key risk is the `AllExciseTaxesPostedForItemLedgerEntry` logic: if an `ItemExciseTax` row is deleted after journal lines are generated but before they are posted, the function would not find that tax type and could mark the ledger entry as fully posted prematurely. This is an unlikely operational sequence but worth keeping in mind. The upgrade codeunit correctly guards with `HasUpgradeTag` / `SetUpgradeTag` and is safe to run multiple times. The API page (`Item Excise Tax API`) has `DeleteAllowed = false` but leaves insert and modify open, which is appropriate for a management API but should be confirmed against the intended access model. Old Item fields remain readable through `CLEAN29` so any extensions referencing them will compile without change until that version.

**Necessity:** The change is necessary. Restricting items to a single excise tax type is a real limitation when an item is subject to both a weight-based and a content-based excise tax simultaneously. The scope — new table, upgrade, obsolescence, UI action, API page — is proportionate to the feature.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=1 system=github pr=9725 round=1 by=alexei-dobriansky at=2026-08-10T08:30:00Z lastSha=2e06e6178c27ce0fdd18d57367d897a64b3bf888 reviewKey=682aebd6f3a5133305b9a2175554486d11c3b52edfac10296672753ff4756371 suggestions=S1@1818b5e0,S2@4944f82e

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 3

# **Recommendation: Accept**

<details>
<summary><strong>What this PR does</strong></summary>

The single commit since round 2 (`22fb847`) applies the two suggestions from the 2026-08-10 review. It adds `ExciseTaxType.SetLoadFields(Description)` before the `Get` call in `CreateItemExciseTaxFromItem` (upgrade codeunit), and removes the redundant enabled-check block and its error label from the `Excise Tax Type Code` `OnValidate` trigger in `ItemExciseTax.Table.al`. Both fixes are minimal and correct. The `ExciseTaxType.Get()` call in the table trigger is preserved — only the dead enabled guard is removed — so `Excise Tax Type Description` is still populated on validate. No further correctness issues were found.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Missing `SetLoadFields` before `ExciseTaxType.Get` in upgrade | Addressed | Added in `22fb847`. |
| S2 | Redundant enabled check duplicates the table relation in `ItemExciseTax` `OnValidate` | Addressed | Block and error label removed in `22fb847`; `Get()` call kept for `Description`. |
</details>

<details>
<summary><strong>New observations (commits since round 2)</strong></summary>

*None — the single authored commit only addresses S1 and S2.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** Both changes are isolated clean-up with no behavior change in the happy path. The `SetLoadFields` addition in the upgrade codeunit reduces read cost but does not change the outcome. The removal of the enabled check in the table trigger has no functional impact because the `TableRelation` already filters to enabled types. No new regression surface was introduced.

**Necessity:** Unchanged from prior rounds — the multiple-excise-taxes-per-item feature is valid and important, and the fix to the posted-state logic was established in round 2. These final tidy-up commits do not affect necessity.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=1 system=github pr=9725 round=3 by=alexei-dobriansky at=2026-08-11T10:45:00Z lastSha=22fb84744e0ce92dffc5d60a27cb8aa47d6ebf2a reviewKey=8a722ccaf101a537560c6f30fb0ede6d68eca7d9bd3f41c0cac65d4c6d11643a suggestions=S1@1818b5e0:addressed,S2@4944f82e:addressed parentRound=2

**Review by alexei-dobriansky — APPROVED** [review]

**Review by djukicmilica — APPROVED** [review]
